eyelid-match-ops/src/primitives/poly.rs

```diff
@@ -130,15 +132,31 @@ pub fn extended_gcd<const MAX_POLY_DEGREE: usize>(
 
 /// This sampling is similar to what will be necessary for YASHE KeyGen.
 /// The purpose is to obtain a polynomial with small random coefficients.
-pub fn sample<const MAX_POLY_DEGREE: usize>(mut rng: ThreadRng) -> Poly<MAX_POLY_DEGREE> {
+pub fn sample_gaussian<const MAX_POLY_DEGREE: usize>(mut rng: ThreadRng) -> Poly<MAX_POLY_DEGREE> {
     let mut res = Poly::zero();
     // TODO: assert that this is less than the modulus of the coefficient
-    let max_coeff = 6;
-    let t = 2;
     for i in 0..MAX_POLY_DEGREE {
-        let coeff_rand = rng.gen_range(1..max_coeff);
-        res[i] = Coeff::from(t * coeff_rand);
+        // TODO: use delta (YASHE param) instead of 3.2
+        // Then those sampling functions should be moved
+        let normal = Normal::new(0.0, 3.2).unwrap();
+        let v: f64 = normal.sample(&mut rng);
+        res[i] = Coeff::from(v as i64);
```

Comment:
These might be maths questions for @bgillesp as well?

Reply:
Most important point: if the sampling is working correctly, integer overflow will not be a problem because the float values being sampled from the normal distribution should be small, i.e. at most one or two digits small, with high probability.  The probabilities decay exponentially with the magnitude of the output value.

Perhaps a radical suggestion, but it actually might make more sense for readability to use a _smaller_ datatype like an `i8` or `i16` for the float to integer conversion, to highlight the fact that the sampled values are expected to be small, and to make it clear that the conversion is not trying to faithfully represent a general `f64`.  In any case we should definitely be sure that the numbers sampled are distributed properly, which could either be its own check, or a test that the overall distribution of coefficients that comes out of the keygen procedure resembles the expected distribution.

Also wanted to mention, it looks like the casting from float to integer types in modern Rust (after 1.45.0) does essentially a clamp operation on values that are out of range -- but it turns out this was a compiler soundness issue until about five years ago: https://blog.rust-lang.org/2020/07/16/Rust-1.45.0.html#fixing-unsoundness-in-casts

Good question on the rounding mode; I think the detail of how the continuous normally distributed samples are to be converted to discrete integer coefficients is discussed in various places in the literature where key generation is addressed, but @emmorais is more familiar with the specific system he's implementing.  We should definitely double-check that the method implemented here matches the spec.

It might also be worth verifying that `Coeff::from` correctly handles signed integer values, but I would be surprised if this wasn't the case.

Reply:
Yes, @bgillesp is right, we don't need big floats, on contrary, their size are limited, thus we could even use smaller types.

Reply:
But the sampling indeed must be implemented carefully. I opened an [issue](https://github.com/Inversed-Tech/eyelid/issues/70) to review the distribution implementation, because it is a task that requires time and a different mindset. Since it can be done in parallel, we can proceed with the rest of the scheme implementation.

Reply:
> Perhaps a radical suggestion, but it actually might make more sense for readability to use a _smaller_ datatype like an `i8` or `i16` for the float to integer conversion, to highlight the fact that the sampled values are expected to be small, and to make it clear that the conversion is not trying to faithfully represent a general `f64`. In any case we should definitely be sure that the numbers sampled are distributed properly, which could either be its own check, or a test that the overall distribution of coefficients that comes out of the keygen procedure resembles the expected distribution.

I would be fine with any type `u32` or smaller. But whatever the we choose, we should assert that the standard deviation is small enough to produce secure keys:
https://github.com/Inversed-Tech/eyelid/issues/70#issuecomment-2083687600

There’s also a significant probability edge case in generation, that we need to confirm isn’t discarding all the random bits (or if it does, the result is out of range and clamped):
https://github.com/Inversed-Tech/eyelid/issues/70#issuecomment-2083732955

This might also tighten the constraint on the standard deviation.

We can deal with this all in the ticket, but it seems important enough that I’ve added it to the milestone as a required security review.